File: auv/scripting/tools/fake_gemini.py

# Standard Library
import argparse
import threading
import copy
import math
import time

# CAUV
import cauv.messaging as msg
import cauv.node as node 
from cauv.debug import debug, warning, error, info
from utils import coordinates

# Third Party
import Image # PIL: MIT-like license
import numpy as np # BSD-type license

Default_Env_File = '/Users/james/Documents/IIB/Project/test_environment.png'
Default_Coordinates_Datum = coordinates.Simulation_Datum

def wrapAngleIntoPlusOrMinusPi(angle):
    while angle > math.pi:
        angle = angle - math.pi*2
    while angle < -math.pi:
        angle = angle + math.pi*2
    return angle

# Image origin is BOTTOM LEFT, rows (y) increase upwards (North), columns (x) increase
# to the right (East)
class Environment(object):
    def __init__(self, image_filename, origin_at_px_x_y, m_per_px, datum):
        self.image = Image.open(image_filename)
        # latitude/longitude position of world xyz coordinate system origin:
        self.datum = datum
        # origin in the image of the world xyz coordinate system origin:
        self.origin = origin_at_px_x_y
        self.m_per_px = float(m_per_px)

        debug('environment is %sx%s metres' %
            (self.image.size[0] * self.m_per_px,
             self.image.size[1] * self.m_per_px)
        )

    def llaToImageCoords(self, latlong):
        northeast = latlong.differenceInMetres(self.datum)
        x = northeast.east
        y = northeast.north
        return self.xyToImageCoords((x,y))

    def xyToImageCoords(self, xy):
        # xy in the same coordinate system as the image (y upwards, x right,
        # but in metres rather than pixels, and with a different origin)
        x = xy[0]; y = xy[1]
        return np.array((x / self.m_per_px - self.origin[0], y / self.m_per_px - self.origin[1]))
    
    def _boundingRect(self, pt_sequence):
        min = list(pt_sequence[0])
        max = list(pt_sequence[0])
        for pt in pt_sequence:
            if pt[0] < min[0]: min[0] = pt[0]
            if pt[0] > max[0]: max[0] = pt[0]
            if pt[1] < min[1]: min[1] = pt[1]
            if pt[1] > max[1]: max[1] = pt[1]
        return (min, max)
    
    def _getLineImageCoords(self, xy, start_angle, end_angle, length, res):
        # x,y:    in pixel-units from pixel origin
        # angle start/end:  measured in radians ccw from x-axis (note that a line at 90
        #         degrees points *negative* y in this system)
        # length: in px
        r = []
        assert(start_angle < end_angle)        
        start_angle = wrapAngleIntoPlusOrMinusPi(start_angle)
        end_angle = wrapAngleIntoPlusOrMinusPi(end_angle)
        
        def angleIsInRage(angle):
            if start_angle <= end_angle:
                return angle >= start_angle and angle <= end_angle
            else:
                return angle <= start_angle or angle >= end_angle

        start_direction = np.array((math.cos(start_angle), math.sin(start_angle)))
        end_direction   = np.array((math.cos(end_angle), math.sin(end_angle)))

        def isPxInImage(x,y):
            return x >= 0 and y > 0 and x < self.image.size[0] and y <= self.image.size[1]
        
        # these are significant optimisations!
        xy_x = xy[0]
        xy_y = xy[1]
        image_sizey = self.image.size[1]
        atan2 = math.atan2
        getpixel = self.image.getpixel

        for bin in xrange(0,res):
            support_px = [0.0]
            value = [0.0]
            #   
            #     - _ _
            #    |       - _ 
            #    |           - 
            # c4 |' - -        /
            #    |      ' .  /
            #    |__       / c2
            # c3 |  ''.. /
            #    |     / c1
            #    |   /
            #    | /
            #    o
            #
            #    y
            #    |
            #    o--> x
            #
            rmin = bin * float(length)/res
            rmax = (bin+1) * float(length)/res
            rmin_sq = rmin**2
            rmax_sq = rmax**2
            c1 = xy + start_direction * rmin
            c2 = xy + start_direction * rmax
            c3 = xy + end_direction * rmin
            c4 = xy + end_direction * rmax
            bbox_min, bbox_max = self._boundingRect((c1,c2,c3,c4))
            #print 'corners:', c1, c2, c3, c4
            ##print 'bbox:', bbox_min, bbox_max

            def pushPx(x, xlocal, xlocal_2, y):
                ylocal = y - xy_y
                r_sq = xlocal_2 + ylocal**2
                if r_sq >= rmin_sq and r_sq <= rmax_sq:
                    xy_angle = atan2(ylocal,xlocal)
                    if angleIsInRage(xy_angle):
                        support_px[0] += 1
                        # convert y into PIL coordinates: origin top
                        value[0] += getpixel(
                             (x, image_sizey - y)
                        )[0]
            
            #import dis
            #print dis.dis(pushPx)
            #sys.exit(0)
            
            for x in xrange(int(math.floor(bbox_min[0])),int(math.ceil(bbox_max[0]))):
                # more micro-optimisations... do I regret not writing this in c++? yes.
                xlocal = x - xy_x
                xlocal_2 = xlocal**2
                map(lambda y: pushPx(x, xlocal, xlocal_2, y),
                    filter(lambda y: isPxInImage(x,y),
                        xrange(int(math.floor(bbox_min[1])),int(math.ceil(bbox_max[1])))
                    )
                )

            if support_px[0] > 0:
                r.append(int(round(value[0]/support_px[0])))
            else:
                # no support? just sample the closest pixel:
                xy_mid = (c1 + c2 + c3 + c4) / 4
                x = round(xy_mid[0])
                y = round(xy_mid[1])
                if isPxInImage(x, y):
                    # convert y into PIL coordinates: origin top
                    r.append(getpixel((x, image_sizey - y))[0])
                else:
                    r.append(0)
        return r


    def getLine(self, latlong, bearing_start, bearing_end, length, res):
        # bearing start/end from north of line centre, degrees
        # length in real world metres
        # res in number-of-pixels
        image_xy = self.llaToImageCoords(latlong)
        angle_end = math.radians(90-bearing_start)
        angle_start = math.radians(90-bearing_end)
        return self._getLineImageCoords(
            image_xy, angle_start, angle_end, length / self.m_per_px, res
        )


class FakeGemini(msg.MessageObserver):
    def __init__(self, node, environment):
        msg.MessageObserver.__init__(self)
        self.node = node
        self.env = environment
        
        self._position_lock = threading.Lock()
        self._position = None

        self._config_lock = threading.Lock()
        self._config = None

        node.addObserver(self)
        node.subMessage(msg.SimPositionMessage())
        node.subMessage(msg.GeminiControlMessage())

    def position(self):
        self._position_lock.acquire()
        r = self._position
        self._position_lock.release()
        return r

    def setPosition(self, sim_position_message):
        self._position_lock.acquire()
        self._position = copy.deepcopy(sim_position_message)
        self._position_lock.release()

    def config(self):
        self._config_lock.acquire()
        r = self._config
        self._config_lock.release()
        return r

    def setConfig(self, gemini_control_message):
        self._config_lock.acquire()
        self._config = gemini_control_message
        self._config_lock.release()
    
    def gemBeamBearings(self):
        r = []
        for i in xrange(1,257):
            r.append(math.degrees(math.asin(((2.0 * i - 256) / 256.0) * 0.86602540)))
        return r

    def onSimPositionMessage(self, m):
        self.setPosition(m)
    
    def onGeminiControlMessage(self, m):
        self.setConfig(m)
        print m

    def mainLoop(self):
        while True:
            config = self.config()
            if not config:
                debug('no gemini config yet!')
                time.sleep(1)
                continue
            time.sleep(config.interPingPeriod)
            position = self.position()
            if not position:
                debug('no sim position yet!')
                continue

            #struct PolarImage
            #{
            #    data: list<byte>;
            #    // currently only raw uint8_t (isodistant responses in fastest varying index) is supported
            #    encoding: ImageEncodingType;
            #    // each bearing in range (-6400 to 6400) * 0x10000, these are the edges of
            #    // the bins that each iso-bearing beam of data falls into, so there are
            #    // num_beams + 1 values. The first bin is bearing_bins[0] --
            #    // bearing_bins[1], and so on.
            #    // Note that these are *bearings* not angles (ie, clockwise from forwards, not
            #    // anticlockwise from x axis (wherever that is...))
            #    bearing_bins: list<int32>;
            #    // number of bins = round(rangeEnd-rangeStart) / rangeConversion)
            #    rangeStart: float;
            #    rangeEnd: float;
            #    // rangeConversion = length of 1 bin in metres
            #    rangeConversion: float;
            #    // time of the ping from which the image is acquired
            #    timeStamp : TimeStamp;
            #}
            #enum SonarID: int8 { Seasprite = 1, Gemini = 2 }            
            #message SonarImage : 31
            #{
            #    source : SonarID;
            #    image : PolarImage;
            #}

            latlong = coordinates.LLACoord(position.latitude, position.longitude, position.altitude)

            beam_bearings = self.gemBeamBearings()

            bearing_bins = msg.floatVec()
            for i, a in enumerate(beam_bearings):
                if i == len(beam_bearings)-1:
                    prev_a = beam_bearings[i-1]
                    bearing_bins.append(a+(a-prev_a)/2)
                    break
                next_a = beam_bearings[i+1]
                if i == 0:
                    bearing_bins.append(a-(next_a-a)/2)
                bearing_bins.append(a+(next_a-a)/2)
            
            print 'at:', latlong, ' = pixels:', self.env.llaToImageCoords(latlong), 'look bearing:', position.orientation.yaw

            beams = []
            for i, a in enumerate(bearing_bins): 
                if i == len(bearing_bins)-1:
                    break
                beam_start = a
                beam_end = bearing_bins[i+1]
                assert(beam_end > beam_start)
                beams.append(self.env.getLine(
                    latlong,
                    position.orientation.yaw + beam_start,
                    position.orientation.yaw + beam_end,
                    config.range, config.rangeLines
                ))
            
            # oops, need to swap order...
            polar_data = msg.byteVec()
            for i in xrange(0, len(beams[0])):
                map(lambda x: polar_data.append(x[i]), beams)
            
            # map bearings from floating point bearing into the on-the-wire
            # integer format (see messages.msg)
            def mapTo64kRange(v):
                r = msg.int32Vec()
                for val in v:
                    i = round(val * 6400 * 0x10000 / 360.0);
                    if i >= 0:
                        i = i % (6400*0x10000);
                    else:
                        i = -((-i) % (6400*0x10000))
                    r.append(int(i))
                return r

            img = msg.PolarImage(
                polar_data,
                msg.ImageEncodingType.RAW_uint8_1,
                mapTo64kRange(bearing_bins),
                0.0,
                config.range,
                config.range / float(config.rangeLines),
                msg.now()
            )

            assert(round((config.range) / (config.range / float(config.rangeLines))) == len(beams[0]))

            self.node.send(msg.SonarImageMessage(msg.SonarID.Gemini, img))
            print '.'


if __name__ == '__main__':
    p = argparse.ArgumentParser()
    p.add_argument('-e', '--environment', dest='env_file', default=Default_Env_File,
            help='Greyscale image to use as environment.')
    p.add_argument('-s', '--environment-scale', dest='env_scale', type=float, default=0.01,
            help='Metres per pixel in the environment file.')
    p.add_argument('-x', dest='env_x', type=float, default=0.0,
            help='x-origin of environment (px offset from top left)')
    p.add_argument('-y', dest='env_y', type=float, default=0.0,
            help='y-origin of environment (px offset from top left)')
    p.add_argument('-p', '--profile', dest='profile', default=False, action='store_true')

    # add a command line argument for this later...
    datum = Default_Coordinates_Datum
    
    opts, args = p.parse_known_args()
    cauv_node = node.Node("py-gem", args)
    
    try:
        e = Environment(opts.env_file, (opts.env_x, opts.env_y), opts.env_scale, datum)
        f = FakeGemini(cauv_node, e)
        if opts.profile:
            import cProfile as profile
            profile.run('f.mainLoop()', 'fakegemini.profile')
        else:
            f.mainLoop()
    finally:
        cauv_node.stop()

File: auv/gemini/fake_gemini.cpp
/* Copyright 2012 Cambridge Hydronautics Ltd.
 *
 * Cambridge Hydronautics Ltd. licenses this software to the CAUV student
 * society for all purposes other than publication of this source code.
 *
 * See license.txt for details.
 *
 * Please direct queries to the officers of Cambridge Hydronautics:
 *     James Crosby    james@camhydro.co.uk
 *     Andy Pritchard   andy@camhydro.co.uk
 *     Leszek Swirski leszek@camhydro.co.uk
 *     Hugo Vincent     hugo@camhydro.co.uk
 */

#include <string>
#include <vector>
#include <stdexcept>

#include <boost/make_shared.hpp>
#include <boost/program_options.hpp>
#include <boost/ref.hpp>

#include <Eigen/Dense>

#include <common/cauv_node.h>
#include <common/msg_classes/wgs84_coord.h>
#include <common/msg_classes/north_east_depth.h>

#include <generated/message_observers.h>
#include <generated/types/GeminiControlMessage.h>
#include <generated/types/SimPositionMessage.h>
#include <generated/types/SonarImageMessage.h>

#include <utility/coordinates.h>
#include <utility/time.h>
#include <utility/rounding.h>
#include <utility/math.h>

#include <debug/cauv_debug.h>

#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>

// !!!!!!! Note that this is a temporary measure: sonar simulation should be
//         done in the main sim using the depth buffer and other magic, which
//         would be a million times more efficient than this.

using namespace cauv;
namespace po = boost::program_options;

// Image origin is BOTTOM LEFT, rows (y) increase upwards (North), columns (x) increase
// to the right (East)
class Environment{
    public:
        Environment(
            std::string const& env_file,
            float origin_x_px,
            float origin_y_px,
            float metres_per_px,
            WGS84Coord const& datum
        ) : m_image(),
            m_datum(datum),
            m_px_origin(origin_x_px, origin_y_px),
            m_metres_per_px(metres_per_px)
        { 
            m_image = cv::imread(env_file.c_str()); 
            if(m_image.channels() == 3){
                cv::Mat tmp;
                cv::cvtColor(m_image, tmp, CV_RGB2GRAY);
                m_image = tmp;
            }
            if(m_image.channels() != 1 || m_image.depth() != CV_8U)
                throw std::runtime_error("use a single-channel 8-bit image!");
            debug() << "environment size is"
                    << m_image.rows * m_metres_per_px << "x"
                    << m_image.cols * m_metres_per_px;
        }
    
    Eigen::Vector2f xyToImgCoords(Eigen::Vector2f const& xy) const{
        return (xy / m_metres_per_px) - m_px_origin;
    }

    Eigen::Vector2f latLongToImgCoords(WGS84Coord const& ll) const{
        const NorthEastDepth ned = ll - m_datum;
        return xyToImgCoords(Eigen::Vector2f(ned.east(), ned.north()));
    }

    std::vector<uint8_t> getLine(
        WGS84Coord const& ll, float bearing_start, float bearing_end, float length, int res
    ) const{
        // bearing start/end from north of line centre, degrees
        // length in real world metres
        // res in number-of-pixels
        const Eigen::Vector2f image_xy = latLongToImgCoords(ll);
        const float angle_end = radians(90-bearing_start);
        const float angle_start = radians(90-bearing_end);
        return getLineImageCoords(
            image_xy, angle_start, angle_end, length / m_metres_per_px, res
        );
    }

    std::vector<uint8_t> getLineImageCoords(
        Eigen::Vector2f const& xy, float start_angle, float end_angle, float length, int res
    ) const{
        // xy:    in pixel-units from pixel origin
        // angle start/end:  measured in radians ccw from x-axis (note that a line at 90
        //         degrees points *negative* y in this system)
        // length: in px
        std::vector<uint8_t> r(res, 0);

        assert(start_angle < end_angle);
        start_angle = wrapAngleIntoPlusOrMinusPi(start_angle);
        end_angle = wrapAngleIntoPlusOrMinusPi(end_angle);
        
        const Eigen::Vector2f start_direction(std::cos(start_angle), std::sin(start_angle));
        const Eigen::Vector2f end_direction(std::cos(end_angle), std::sin(end_angle));
        
        const int image_sizex = m_image.cols;
        const int image_sizey = m_image.rows;
        
        Eigen::Vector2f c1 = xy;
        Eigen::Vector2f c2 = xy;
        Eigen::Vector2f c3 = xy;
        Eigen::Vector2f c4 = xy;
        float rmin = 0;
        float rmin_sq = 0;
        float rmax = 0;
        float rmax_sq = 0;
        for(int bin = 0; bin < res; bin++){
            float support_px = 0.0;
            float value = 0.0;
            //   
            //     - _ _
            //    |       - _ 
            //    |           - 
            // c4 |' - -        /
            //    |      ' .  /
            //    |__       / c2
            // c3 |  ''.. /
            //    |     / c1
            //    |   /
            //    | /
            //    o
            //
            //    y
            //    |
            //    o--> x
            //
            rmin = rmax;
            rmax = (bin+1) * float(length)/res;
            rmin_sq = rmax_sq;
            rmax_sq = rmax * rmax;
            c1 = c2;
            c2 = xy + start_direction * rmax;
            c3 = c4;
            c4 = xy + end_direction * rmax;
            const Rect bbox = boundingRect(c1, c2, c3, c4);
            
            for(int x = bbox.min[0]; x < bbox.max[0]+1; x++){
                const float xlocal = x - xy[0];
                for(int y = bbox.min[1]; y < bbox.max[1]+1; y++){ 
                    if(x >= 0 && y > 0 && x < image_sizex && y <= image_sizey){
                        const float ylocal = y - xy[1];
                        const float r_sq = xlocal*xlocal + ylocal*ylocal;
                        if(r_sq >= rmin_sq && r_sq <= rmax_sq){
                            const float xy_angle = std::atan2(ylocal, xlocal);
                            if(angleIsInRange(start_angle, xy_angle, end_angle)){
                                support_px += 1;
                                // convert y into Image coordinates: origin top
                                value += m_image.at<uint8_t>(image_sizey - y,x);
                            }
                        }
                    }
                }
            }

            if(support_px > 0){
                r[bin] = uint8_t(0.5 + value / support_px);
            }else{
                // no support? just sample the closest pixel:
                Eigen::Vector2f xy_mid = (c1 + c2 + c3 + c4) / 4;
                const int x = int(0.5 + xy_mid[0]);
                const int y = int(0.5 + xy_mid[1]);
                if(x >= 0 && y > 0 && x < image_sizex && y <= image_sizey){
                    // convert y into Image coordinates: origin top
                    r[bin] = m_image.at<uint8_t>(image_sizey - y,x);
                }
            }
        }
        return r;
    }

    private:
        static float wrapAngleIntoPlusOrMinusPi(float angle){
            while(angle > M_PI)
                angle -= M_PI*2;
            while(angle < -M_PI)
                angle += M_PI*2;
            return angle;
        }
        
        struct Rect{
            Eigen::Vector2f min;
            Eigen::Vector2f max;
        };
        static Rect boundingRect(
            Eigen::Vector2f const& a,
            Eigen::Vector2f const& b,
            Eigen::Vector2f const& c,
            Eigen::Vector2f const& d
        ){
            Rect r = {a, a};
            Eigen::Vector2f min = a;
            Eigen::Vector2f max = a;
            for(int i = 0; i < 2; i++){
                if(b[i] < r.min[i]) r.min[i] = b[i];
                if(b[i] > r.max[i]) r.max[i] = b[i];
                if(c[i] < r.min[i]) r.min[i] = c[i];
                if(c[i] > r.max[i]) r.max[i] = c[i];
                if(d[i] < r.min[i]) r.min[i] = d[i];
                if(d[i] > r.max[i]) r.max[i] = d[i];
            }
            return r;
        }

        static bool angleIsInRange(float start_angle, float angle, float end_angle){
            if(start_angle <= end_angle)
                return angle >= start_angle && angle <= end_angle;
            else
                return angle <= start_angle || angle >= end_angle;
        }

        cv::Mat m_image;
        WGS84Coord m_datum;
        Eigen::Vector2f m_px_origin;
        float m_metres_per_px;
};

class FakeGemini: public MessageObserver{
    public:
        FakeGemini(
            Environment const& env,
            CauvNode& node
        ) : MessageObserver(),
            m_environment(env),
            m_node(node),
            m_inited(0),
            m_position(),
            m_gemcontrol(){
            
        }

        void start(){
            mainLoop();
        }

        virtual void onSimPositionMessage(SimPositionMessage_ptr p){
            m_inited |= GotSimPos;
            m_position = p;
        }

        virtual void onGeminiControlMessage(GeminiControlMessage_ptr p){
            m_inited |= GotGemControl;
            m_gemcontrol = p;
        }

        void mainLoop() const{
            for(;;){
                if(!(m_inited & GotGemControl)){
                    debug() << "no gemini control messages received";
                    msleep(1000);
                    continue;
                }
                const GeminiControlMessage_ptr gemcontrol = m_gemcontrol; 
                msleep(gemcontrol->interPingPeriod() * 1000);

                if(!(m_inited & GotSimPos)){
                    debug() << "no sim position messages received";
                    continue;
                }
                SimPositionMessage_ptr p = m_position;
                const WGS84Coord pos = p->location(); 
                const float yaw = p->orientation().yaw;
                const TimeStamp ts = now();
                
                const static std::vector<float> bearing_bins = bearingBinDegrees();
                const static std::vector<int> bearing_bins_otw = bearingBinsOTW();
                
                const float range = gemcontrol->range();
                const float rangeLines = gemcontrol->rangeLines();
                const int num_beams = 256;

                debug() << "at:" << pos
                        << " =pixels:" << m_environment.latLongToImgCoords(pos).transpose()
                        << " look bearing:" << yaw;

                std::vector<uint8_t> beams;
                beams.reserve(num_beams*rangeLines);

                for(std::size_t i = 0; i+1 < bearing_bins.size(); i++){
                    const float beam_start = bearing_bins[i] + yaw;
                    const float beam_end = bearing_bins[i+1] + yaw;
                    assert(beam_end > beam_start);
                    const std::vector<uint8_t> beam = m_environment.getLine(
                        pos, beam_start, beam_end, range, rangeLines
                    );
                    beams.insert(beams.end(), beam.begin(), beam.end());
                }

                // oops, need to swap order...
                std::vector<uint8_t> beams_swapped;
                beams_swapped.reserve(beams.size());
                for(int i = 0; i < rangeLines; i++)
                    for(int b = 0; b < num_beams; b++)
                        beams_swapped.push_back(beams[b * rangeLines + i]);
                
                m_node.send(boost::make_shared<SonarImageMessage>(
                    SonarID::Gemini, PolarImage(
                        beams_swapped,
                        ImageEncodingType::RAW_uint8_1,
                        bearing_bins_otw,
                        0.0,
                        range,
                        range / rangeLines,
                        ts
                    )
                ));
            }
        }

    private:
        static std::vector<float>  bearingBinDegrees(){
            std::vector<float> r(257);
            for(uint32_t beam = 0; beam <= 256; beam++){
                // see Gemini Interface Specification Appendix B
                double radians = std::asin(((2*(beam+0.5) - 256) / 256) * 0.86602540);
                r[beam] = radians * (180.0 / M_PI);
            }
            return r;
        }
        static std::vector<int32_t> bearingBinsOTW(uint32_t num_beams=256){
            std::vector<int32_t> r;
            r.reserve(num_beams+2);
            const double radConvert = (180.0 / M_PI) * (6400.0/360.0) * 0x10000;
            for(uint32_t beam = 0; beam <= num_beams; beam++){
                // see Gemini Interface Specification Appendix B
                double radians = std::asin(((2*(beam+0.5) - num_beams) / num_beams) * 0.86602540);
                r.push_back(round(radConvert * radians));
            }
            return r;
        }


        Environment m_environment;
        CauvNode& m_node;

        enum {GotSimPos = 1, GotGemControl = 2};
        int m_inited;

        SimPositionMessage_ptr   m_position;
        GeminiControlMessage_ptr m_gemcontrol;
};


class FakeGeminiNode: public CauvNode{
    public:
        FakeGeminiNode() : CauvNode("FakeGem"){ }
    protected:
        virtual void onRun(){
            if(!m_sim){
                error() << "not initialised";
                return;
            }

            // required message subscriptions:
            subMessage(GeminiControlMessage());
            subMessage(SimPositionMessage());

            addMessageObserver(m_sim);
            m_sim->start();
        }
        virtual void addOptions(po::options_description& desc,
                                po::positional_options_description& pos){
            CauvNode::addOptions(desc, pos);

            desc.add_options()
                ("environment,e", po::value<std::string>()->required(), "The environment file (single-channel image)")
                ("scale,s", po::value<float>()->required(), "The environment scale (metres per pixel)")
                ("origin-x,x", po::value<float>(), "x-origin of environment (px from top-left)")
                ("origin-y,y", po::value<float>(), "y-origin of environment (px from top-left)")
            ;

        }
        virtual int useOptionsMap(po::variables_map& vm,
                                  po::options_description& desc){
            int ret = CauvNode::useOptionsMap(vm, desc);
            if(ret) return ret;
            
            const std::string env_file = vm["environment"].as<std::string>();
            const float scale = vm["scale"].as<float>();
            float origin_x_px = 0;
            float origin_y_px = 0;
            if(vm.count("origin-x")) origin_x_px = vm["origin-x"].as<float>();
            if(vm.count("origin-y")) origin_y_px = vm["origin-y"].as<float>();

            m_sim = boost::make_shared<FakeGemini>(Environment(
                env_file, origin_x_px, origin_y_px, scale, WGS84Coord::fromDatum("river cam")
            ), boost::ref(*this));

            return 0;
        }
    private:
        boost::shared_ptr<FakeGemini> m_sim;
};


static FakeGeminiNode* node;

void cleanup(){
    info() << "Cleaning up...";
    CauvNode* oldnode = node;
    node = 0;
    delete oldnode;
    info() << "Clean up done.";
}

void interrupt(int sig){
    std::cout << std::endl;
    info() << BashColour::Red << "Interrupt caught!";
    cleanup();
    signal(SIGINT, SIG_DFL);
    raise(sig);
}

int main(int argc, char** argv){
    signal(SIGINT, interrupt);
    node = new FakeGeminiNode();
    try{
        if(node->parseOptions(argc, argv) == 0)
            node->run(false);
    }catch(boost::program_options::error& e){
        error() << e.what();
    }
    cleanup();
    return 0;
}
